Fix error backtrace formatting on Ruby 3.4+

orien opened the pull request:
# Description

## Context

Cucumber performs filtering and formatting on error backtraces before displaying them in terminal output.

When running on Ruby 3.4, the backtrace filtering and formatting is not working. This is defect is caused by the new backtrace format introduced in Ruby 3.4.

See:
- https://www.ruby-lang.org/en/news/2024/12/25/ruby-3-4-0-released/
- https://bugs.ruby-lang.org/issues/16495
- https://bugs.ruby-lang.org/issues/19117

## Change

Update the `Cucumber::Glue::InvokeInWorld` and `Cucumber::Formatter::BacktraceFilter` classes to support both styles of backtraces.

Fixes #1770

## Type of change
- Bug fix (non-breaking change which fixes an issue)

# Checklist:

Your PR is ready for review once the following checklist is
complete. You can also add some checks if you want to.

- [ ] Tests have been added for any changes to behaviour of the code
- [x] New and existing tests are passing locally and on CI
- [x] `bundle exec rubocop` reports no offenses
- [ ] RDoc comments have been updated
- [x] CHANGELOG.md has been updated

orien commented on .github/workflows/test.yaml:
```diff
@@ -17,12 +17,15 @@ jobs:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        ruby: ['3.0', '3.1', '3.2', '3.3']
+        ruby: ['3.0', '3.1', '3.2', '3.3', '3.4']
```
Add Ruby 3.4 to the CI build matrix.

orien commented on compatibility/cck_spec.rb:
```diff
@@ -9,7 +9,7 @@
 
   CCK::Examples.gherkin.each do |example_name|
-    describe "'#{example_name}' example" do
+    describe "'#{example_name}' example", pending: example_name == 'empty' do # TODO: resolve the empty example.
```
Some tests were failing on the `main` branch. Specifically, the CCK specs on the `empty` example.

Let's mark these specs as pending until someone has time to investigate and resolve the issue.

luke-hill replied:
Happy for this as this is a major bump for cucumber ruby to v10. But this is something we should fix soon. I'll make this a issue on GH

orien replied:
These tests are passing now. I've reverted this change.

voxik commented:
Trying these fixes to make [this](https://github.com/cucumber/cucumber-ruby-wire/blob/f5879f379660b7146de391e043e92021306920b6/features/handle_unexpected_response.feature#L30) cucumber-wire test pass. Unfortunately without success. Is it something what could be fixed on Cucumber side? Or does the cucumber-wire need medicine like this:

~~~diff
diff --git a/features/handle_unexpected_response.feature b/features/handle_unexpected_response.feature
index 2d7747e..4e0aa8c 100644
--- a/features/handle_unexpected_response.feature
+++ b/features/handle_unexpected_response.feature
@@ -25,7 +27,7 @@ Feature: Handle unexpected response
       | ["begin_scenario"]                                   | ["yikes"]                           |
       | ["step_matches",{"name_to_match":"we're all wired"}] | ["success",[{"id":"1", "args":[]}]] |
     When I run `cucumber -f pretty`
-    Then the output should contain:
+    Then the output should match:
       """
-      undefined method `handle_yikes'
+      undefined method [`']handle_yikes'
       """
~~~

luke-hill reviewed (commented):
Partially reviewed, probably won't have time this week. But this is near the top of my list.

Great work finding a good replacement!

orien commented on spec/cucumber/glue/proto_world_spec.rb:
```diff
@@ -72,8 +72,9 @@ module Glue
 
-          it 'attached the styring version on the object' do
-            expect(@out.string).to include '{:a=>1, :b=>2, :c=>3}'
+          it 'attached the string version on the object' do
```
I've updated the test in 117688fa5204abac556fe95056a3b547a89855b5 to more accurately match the test description. Is that preferable?

luke-hill replied:
Sorry I never scrolled up. I can see this is in a heredoc which is why there was no utilised location edited. I'll re-review this when I get time.

luke-hill commented on features/docs/extending_cucumber/custom_formatter.feature:
```diff
@@ -45,7 +45,7 @@ Feature: Custom Formatter
             @io = config.out_stream
             config.on_event :test_run_finished do |event|
-              @io.print options.inspect
+              @io.print options.to_json
```
I'm unresolving this so we know to look here for documentation purposes as/when a v10 is cut.

For now as part of this PR can you add some notes here: https://github.com/cucumber/cucumber-ruby/blob/main/upgrading_notes/10.0.0.md - if you are confident in doing so. Or if you'd prefer I can do it once this PR is merged. I'm flexible.

orien replied:
I'm not sure what you're referring to specifically. This is a test file. The changes here shouldn't affect downstream users.

luke-hill commented on spec/cucumber/glue/proto_world_spec.rb:
```diff
@@ -87,16 +92,15 @@ module Glue
           define_steps do
             When('{word} {word} {word}') do |subject, verb, complement|
-              log "subject: #{subject}", "verb: #{verb}", "complement: #{complement}", subject: subject, verb: verb, complement: complement
+              log "subject: #{subject}", "verb: #{verb}", "complement: #{complement}"
```
Can we change this to be hard-coded. Aruba steps are designed to showcase simple examples and this has 3 parameter types for "I have no idea what reason...." So if this just becomes `When('monkey eats banana')` then we can make the logger step also hard coded and the test showcasing that the logger having 3 items gets translated will be a bit easier to understand

Notwithstanding that, I'm not sure I agree that "is" what happens though. So I'll probably need to dig into this.

orien replied:
Updated in 4740677ec62def308572ae349c1c649991222a04.

luke-hill commented on spec/cucumber/glue/proto_world_spec.rb:
```diff
@@ -68,35 +68,39 @@ module Glue
           end
 
-          it 'logs each parameter independently' do
+          it 'logs each entry independently' do
```
This feels super weird to me. I'll be raising this in the contributor meeting as I'm curious why we ever built it to do this. But yes this is correct.

Can you wrap the include in `()`

luke-hill reviewed (commented):
Fully reviewed this PR now. If you can respond to the points we can look to get this merged in soon and included in v10.

You'll also likely need to do some form of rebase. But it might be worth waiting as I'm doing a few derpy things in `main` currently.

orien commented on lib/cucumber/glue/invoke_in_world.rb:
```diff
@@ -49,7 +52,12 @@ def self.cucumber_compatible_arity?(args, block)
       def self.cucumber_run_with_backtrace_filtering(pseudo_method)
         yield
       rescue Exception => e
-        instance_exec_invocation_line = "#{__FILE__}:#{__LINE__ - 2}:in `cucumber_run_with_backtrace_filtering'"
```
I've updated this a little. Is it clearer now?

luke-hill replied:
I mean that's fine. But it looks like the previous implementation was wrong as here it's now the same for both. I don't mind that being in-lined or not.

orien replied:
Oh, both implementations work and provide a correct result.

Ruby `__LINE__` is a keyword that gets replaced with the line number in the source code.

Previously, `__LINE__` was on line 57 and 59, which gave (57 - 4 = 53) and (59 - 6 = 53) respectively. Now `__LINE__` is on line 55: (55 - 2 = 53). All reference the `yield` line of code.

orien commented:
> Fully reviewed this PR now. If you can respond to the points we can look to get this merged in soon and included in v10.
> 
> You'll also likely need to do some form of rebase. But it might be worth waiting as I'm doing a few derpy things in `main` currently.

Thanks @luke-hill. I think I've addressed everything and rebased the branch against `main`. Please let me know if/when I need to rebase again (if you're still working for v10).

luke-hill commented on lib/cucumber/formatter/backtrace_filter.rb:
```diff
@@ -44,7 +44,7 @@ def exception
         if ::ENV['CUCUMBER_TRUNCATE_OUTPUT']
           # Strip off file locations
           filtered = filtered.map do |line|
-            line =~ /(.*):in `/ ? Regexp.last_match(1) : line
```
Sorry there is a `#match?` boolean method you can use here. This is much more performant and it won't store the matches, just check if it does match

orien replied:
I don't think we can use `#match?`, given we use a regexp group to capture a part of the match.

luke-hill replied:
Tell you what, to makes things easier. I'll make this as a breakout task for me to tackle as tech-debt. I'm not familiar with this file but lots of this I think is either super edge case or possibly partially redundant. I'll make this issue [here](https://github.com/cucumber/cucumber-ruby/issues/1777)

luke-hill reviewed (commented):
3/6 files are marked as good from me, couple of minor things to fix - I'll leave the note up for the doc updates as I still haven't got round to that yet. I will soon though as I've merged my chunky v10 diff.

luke-hill commented:
For now @orien this is in a decent spot. So you can rest up for a while now as there's not much to do.

luke-hill commented:
Hi @orien 

It was a lot of work to do the backport and I'd rather not do it. Furthermore whilst my time in the past month or two was focussed on other items. I'm now focussing on this. So v10 should be out in a matter of weeks. Then hopefully we can just continue working standardly in the v10 realm
